rs/p2p/artifact_downloader/src/fetch_stripped_artifact/assembler.rs

```diff
@@ -222,33 +227,220 @@ async fn get_or_fetch<P: Peers>(
     .await
 }
 
-#[derive(Debug, PartialEq)]
-pub(crate) enum InsertionError {}
+#[derive(Debug, PartialEq, Error)]
+pub(crate) enum InsertionError {
+    #[error("Trying to insert an ingress message which was never missing")]
+    NotNeeded,
+    #[error("Trying to insert an ingress message which was already inserted")]
+    AlreadyInserted,
+}
 
-#[derive(Debug, PartialEq)]
-pub(crate) enum AssemblyError {}
+#[derive(Debug, Error)]
+pub(crate) enum AssemblyError {
+    #[error("The block proposal is missing ingress message with id {0}")]
+    Missing(IngressMessageId),
+    #[error("The block proposal cannot be deserialized {0}")]
+    DeserializationFailed(ProxyDecodeError),
+}
 
 impl StrippedBlockProposal {
     /// Returns the list of [`IngressMessageId`]s which have been stripped from the block.
-    // TODO(kpop): Implement this
     pub(crate) fn missing_ingress_messages(&self) -> Vec<IngressMessageId> {
-        unimplemented!()
+        self.stripped_ingress_payload
+            .ingress_messages
+            .iter()
+            .filter_map(|maybe_ingress| match maybe_ingress {
+                MaybeStrippedIngress::Full(_, _) => None,
+                MaybeStrippedIngress::Stripped(ingress_message_id) => Some(ingress_message_id),
+            })
+            .cloned()
+            .collect()
     }
 
     /// Tries to insert a missing ingress message into the block.
-    // TODO(kpop): Implement this
     pub(crate) fn try_insert_ingress_message(
         &mut self,
-        _ingress_message: SignedIngress,
+        ingress_message: SignedIngress,
     ) -> Result<(), InsertionError> {
-        unimplemented!()
+        let ingress_message_id = IngressMessageId::from(&ingress_message);
+
+        let ingress = self
+            .stripped_ingress_payload
+            .ingress_messages
+            .iter_mut()
+            .find(|ingress| match ingress {
+                MaybeStrippedIngress::Full(id, _) => *id == ingress_message_id,
+                MaybeStrippedIngress::Stripped(id) => *id == ingress_message_id,
+            })
+            .ok_or(InsertionError::NotNeeded)?;
```

Comment:
I think in this case iterating is fine because it is only max 1000 elements so a linear search is likely not slower than building an index. Maybe it is worth adding a comment that iterating is fine since the max number of elements is known.

Reply:
Agreed, done.